### PR TITLE
PluginExtension: Added debug log

### LINKED ISSUE #89476: PluginExtensions: Add a reactive log to make it easier to debug extensions
### Description
When adding/exposing `links` or `components` we have a set of rules that needs to be followed when it comes to the format of the `id`, required fields etc. Lots of people seems to get stuck while trying to add an extension due to some small errors (typings, id missmatch etc.)

Today we print a couple of debug logs in the console but due too a lot of noise they are really easy to miss. Instead we are planning to add a proper log that will persist logs from all interactions with the extension registry when you run Grafana in development mode.

We could then expose that log in the extensions admin view to make it really easy to understand why things might not show up the way the developer expected.

### Todo
- [ ] Introduce a reactive log that will store all log events for a user session (maybe have some kind of upper limit).
- [ ] Write a log item when adding things to the registry (info, warn and error).
- [ ] Write a log item when getting things from the registry.

### Related
https://github.com/grafana/grafana/pull/85944
https://github.com/grafana/grafana/issues/77856

### PR DESCRIPTION
**Pull request by mckn**
**What is this feature?**

This PR adds an admin page that will display all the logs collected by Grafana while developing UI extensions. It will only be available when running Grafana in development mode.

We are using broadcast channels to communicate logs between different tabs/browsers. So you can easily open the extensions log view in one browser/tab and debug your extensions in another tab.

events-log-low-res

**Why do we need this feature?**

Getting started with UI extensions can e a bit painful. This feature will give better insights in what is happening under the hood and why/how Grafana will (not) display your extensions.

**Who is this feature for?**

Developers adding UI extensions to either plugin or Grafana.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/89476

**Special notes for your reviewer:**
Started to add [tests](https://github.com/grafana/grafana/blob/mckn/extensions-logging-2-LV-tests/public/app/features/plugins/extensions/logs/LogViewer.test.tsx) for the LogViewer but ran into some issues to resolve the panels via the scene framework. So I skipped them for now.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: public/app/features/plugins/extensions/logs/VizGrid.tsx [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    display: 'grid',
+    flexGrow: 1,
+    gridTemplateColumns: 'repeat(auto-fit, minmax(400px, 1fr))',
+    gridAutoRows: '320px',
+    columnGap: `8px`,
+    rowGap: `8px`,
+  };
+
+  return <div style={style}>{props.children}</div>;
+}
```
Comment (torkelo) [inline public/app/features/plugins/extensions/logs/VizGrid.tsx:18]: can remove this and use VizGridLayout (was exported in scenes-react 15.16.1

File: public/app/features/plugins/extensions/logs/log.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
+import { isString } from 'lodash';
+import { nanoid } from 'nanoid';
+import { Observable, ReplaySubject } from 'rxjs';
+
+import { LogLevel } from '@grafana/data';
+
+export type LogItem = {
+  level: LogLevel;
+  timestamp: number;
+  labels: Record<string | symbol, unknown>;
```
Comment (sunker) [inline public/app/features/plugins/extensions/logs/log.ts:10]: Can use labels type from grafana/data

**Review by sunker — COMMENTED** [review]
Really nice! 👏  I'm in favour of this pattern using a transformation rather than doing the filtering in the data source. 

From a usability standpoint, it's not ideal that one has to first navigate to the admin page and then reload another tab for messages to appear in the log. I guess this is because there's no subscriber to the channel before the logs viewer is initialized? How about we store the last 100 log items in memory within the loge class, and then post them once a subject has been registered? 

Before merging this PR, I guess it's best to remove all usage of this log within the registries and do that in a follow up pr istead?

File: public/app/features/plugins/extensions/logs/LogViewer.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    data: data,
+  });
+
+  return (
+    <Page
+      navId="extensions"
+      actions={<LogViewFilters provider={data} filteredProvider={filteredData} filter={filter} onChange={setFilter} />}
+    >
+      <VizGrid>
+        <VizPanel title="Logs" viz={logsViz} dataProvider={filteredData} />
```
[the remark this answers is not in the document]
Comment (mckn) [reply public/app/features/plugins/extensions/logs/LogViewer.tsx:55]: It is a required field so I can not remove it.

[the remark this answers is not in the document]
Comment (sunker) [reply public/app/features/plugins/extensions/logs/LogViewer.tsx:58]: Guess you can just pass an empty string?

File: public/app/features/plugins/extensions/logs/filterTransformation.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                }
+
+                if (!isEmpty(filter.extensionPointIds)) {
+                  if (!filter.extensionPointIds?.has(extensionPointId)) {
+                    continue;
+                  }
+                }
+
+                copyRow(frame, target, index);
+                continue;
```
Comment (leventebalogh) [inline public/app/features/plugins/extensions/logs/filterTransformation.ts:95]: Hmm, could we reduce the nesting like this? (If I understand correctly we filter out a row if any of the filters don't match)

```ts
// Plugin id doesn't match
if (
  !isEmpty(filter.pluginIds) && 
  !filter.pluginIds?.has(pluginId)
) {
  continue;
}

// Extension point id doesn't match
if (
  !isEmpty(filter.extensionPointIds) && 
  !filter.extensionPointIds?.has(extensionPointId)
) {
  continue;
}

// Severity doesn't match
if (
  !isEmpty(filter.severity) && 
  !filter.severity?.has(severity)
) {
  continue;
}

copyRow(frame, target, index);
filtered.push(createDataFrame(target));
```

File: public/app/features/plugins/extensions/logs/log.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
+import { isString } from 'lodash';
+import { nanoid } from 'nanoid';
+import { Observable, ReplaySubject } from 'rxjs';
+
+import { Labels, LogLevel } from '@grafana/data';
+
+export type LogItem = {
```
Comment (leventebalogh) [inline public/app/features/plugins/extensions/logs/log.ts:7]: Nit: maybe `ExtensionsLogItem` would be more specific.
```suggestion
export type ExtensionsLogItem = {
```

File: public/app/features/plugins/extensions/registry/AddedComponentsRegistry.ts [hunk 6]
```diff
@@ -57,19 +63,23 @@ export class AddedComponentsRegistry extends Registry<
...
 
         const result = {
           pluginId,
-          component: wrapWithPluginContext(pluginId, config.component),
+          component: wrapWithPluginContext(pluginId, config.component, pointIdLog),
           description: config.description,
           title: config.title,
         };
 
+        pointIdLog.debug(`Added component from '${pluginId}' to '${extensionPointId}'`);
```
Comment (leventebalogh) [inline public/app/features/plugins/extensions/registry/AddedComponentsRegistry.ts:81]: 👏

File: public/app/features/plugins/extensions/logs/LogViewer.tsx
Comment (leventebalogh) [inline public/app/features/plugins/extensions/logs/LogViewer.tsx:1]: Probably would be nice to add some tests for this view.

**Review by leventebalogh — COMMENTED** [review]
Nice! 👏 The use of `scenes-react` made it easier to read in my opinion!

File: public/app/features/plugins/extensions/logs/dataSource.ts
Comment (leventebalogh) [inline public/app/features/plugins/extensions/logs/dataSource.ts:1]: Hmm, probably would be nice to add some tests for the datasource logic as well.

File: public/app/features/plugins/extensions/logs/filterTransformation.ts (hunk 4 above)
Comment (mckn) [reply public/app/features/plugins/extensions/logs/filterTransformation.ts:95]: 🎉

File: public/app/features/plugins/extensions/logs/dataSource.ts
Comment (mckn) [reply public/app/features/plugins/extensions/logs/dataSource.ts:1]: True, I have added a todo to get that in place. Will look at it tomorrow.

File: public/app/features/plugins/extensions/logs/LogViewer.tsx
Comment (mckn) [reply public/app/features/plugins/extensions/logs/LogViewer.tsx:1]: True, I have added a todo to get that in place. Will look at it tomorrow.

**Review by leventebalogh — APPROVED** [review]
LGTM 🚀

**Pull request merged**